File: python/benches/ode.py
from de.utils import bench
from de.ode import ODE, Method, Config


if __name__ == "__main__":
    rk4_config = Config()
    euler_config = Config(method=Method.Euler)
    ode = ODE(
        1.0,
        1.0,
        lambda t: -t,
    )

    def rk4():
        ode.solve(rk4_config)

    def euler():
        ode.solve(euler_config)

    def rk4_numba():
        ode.solve_numba(rk4_config)

    def euler_numba():
        ode.solve_numba(euler_config)

    bench("ode-rk4", rk4)
    bench("ode-euler", euler)
    bench("ode-rk4-numba", rk4_numba)
    bench("ode-euler-numba", euler_numba)

File: rust/benches/ode.rs
use criterion::{Criterion, criterion_group, criterion_main};
use de::ode::{Config, Method, ODE};
use std::hint::black_box;

fn criterion_benchmark(c: &mut Criterion) {
    let ode = ODE {
        lambda: 1.0,
        initial_condition: 1.0,
        source_term: |t| -t,
    };
    let rk4_config = Config::default();
    let euler_config = Config {
        method: Method::Euler,
        ..Default::default()
    };
    c.bench_function("ode-euler", |b| {
        b.iter(|| {
            let _ = ode.solve(black_box(&euler_config));
        })
    });
    c.bench_function("ode-rk4", |b| {
        b.iter(|| {
            let _ = ode.solve(black_box(&rk4_config));
        })
    });
}

criterion_group!(benches, criterion_benchmark);
criterion_main!(benches);
